pkg/remotewrite/client.go

```diff
@@ -50,11 +58,23 @@ func (r *Client) Ingest(ctx context.Context, in *ingestion.IngestInput) error {
 	r.enhanceWithAuth(req)
 
 	req = req.WithContext(ctx)
-	r.log.Debugf("Making request to %s", req.URL.String())
+
+	dump, _ := httputil.DumpRequestOut(req, true)
+	// ignore errors
+	if err == nil {
+		r.metrics.sentBytes.Add(float64(len(dump)))
+	}
```

Comment:
dumping before setting up body + using in.Profile.Bytes() sounds good

Reply:
actually i would prefer to setup the body when creating the request, since the constructor already sets up few things like the ContentLength https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/net/http/request.go;l=893

Reply:
We don't need to consider HTTP overhead, it's almost always negligible comparing to the request body in our case: profiles easily reach a megabyte or two in size. `httputil.DumpRequestOut`, apart from many other steps, allocates a new buffer and copies body there; which is worse, the buffer is of zero size, therefore effective memory allocation may rich 2x of the original size. I don't think this waste is justified. If we needed to keep track of every byte we send via network, I'd suggest implementing a wrapper for `Client.Transport.DialContext` that tracks writes to the connection

Reply:
Yeah, request dumping in hots paths should be avoided at all costs. Getting length of `in.Profile.Bytes()` should be good enough for now, and if we want to get really precise we can do a custom `Client.Transport.DialContext` that would return a wrapped connection. Something like this could work:
```
// Package conncounter provides a net.Conn wrapper that keeps a count of bytes sent in both direction.
package conncounter

import (
	"net"
	"sync/atomic"
)

// Conn is the wrapper that extends net.Conn with byte counting capabilities
type Conn struct {
	net.Conn
	readCount  uint64
	writeCount uint64
}

// NewConn should be used to wrap an existing net.Conn with conncounter.Conn
func NewConn(conn net.Conn) *Conn {
	return &Conn{
		Conn: conn,
	}
}

func (counter *Conn) Read(buf []byte) (int, error) {
	n, err := counter.Conn.Read(buf)
	atomic.AddUint64(&counter.readCount, uint64(n))
	return n, err
}

func (counter *Conn) ReadCount() uint64 {
	return atomic.LoadUint64(&counter.readCount)
}

func (counter *Conn) Write(buf []byte) (int, error) {
	n, err := counter.Conn.Write(buf)
	atomic.AddUint64(&counter.writeCount, uint64(n))
	return n, err
}

func (counter *Conn) WriteCount() uint64 {
	return atomic.LoadUint64(&counter.writeCount)
}
```